lib/plug/cowboy/telemetry.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Plug.Cowboy.Telemetry do
+  def init(stream_id, req, opts) do
+    system_time = System.system_time()
+    start_time = System.monotonic_time()
+
+    :telemetry.execute([:plug_cowboy, :stream_handler, :start], %{system_time: system_time}, %{
+      host: req.host,
+      method: req.method,
+      request_path: req.path,
+      port: req.port,
+      remote_ip: req.peer |> elem(0),
+      scheme: req.scheme,
+      req_headers: req.headers |> Map.to_list(),
+      query_string: req.qs
+    })
+
+    {commands, state} = :cowboy_stream.init(stream_id, req, opts)
+    {commands, [state, start_time]}
+  end
+
+  def info(stream_id, info, [state, start_time]) do
+    end_time = System.monotonic_time()
+
+    case info do
+      {:response, <<status::binary-size(3)>> <> _message, headers, body} ->
+        :telemetry.execute(
+          [:plug_cowboy, :stream_handler, :stop],
+          %{duration: end_time - start_time},
+          %{
+            status: status |> String.to_integer(),
+            resp_headers: headers |> Map.to_list(),
+            body: body
+          }
+        )
+
+      _ ->
+        :ignore
+    end
+
+    :cowboy_stream.info(stream_id, info, state)
+  end
+
+  def info(stream_id, info, state) do
+    case info do
+      {:EXIT, _pid, reason} ->
```

Comment:
When this `info` callback is executed, the state that I stored in `init` is no longer there, otherwise the function head above would get called. That's preventing me from calculating the `duration` here...

I'm not sure what I'm missing because the `cowboy_metrics_h` stream handler _does_ appear to be getting it's state passed to it:

https://github.com/ninenines/cowboy/blob/master/src/cowboy_metrics_h.erl#L187

All I am really trying to do is store a little extra state and have it available in all the callbacks

Reply:
I think I figured it out :)